imapautofiler/secrets.py

```diff
@@ -0,0 +1,79 @@
+import logging
+import getpass
+
+import keyring
+
+LOG = logging.getLogger('imapautofiler.client')
+
+
+class FixedPasswordSecret:
+    def __init__(self, password):
+        self.password = password
+
+    def get_password(self):
+        return self.password
+
+
+class KeyringPasswordSecret:
+
+    def __init__(self, hostname, username):
+        self.hostname = hostname
+        self.username = username
+
+    def get_password(self):
+        password = keyring.get_password(self.hostname, self.username)
+        if not password:
+            password = getpass.getpass(
+                'Password for {} (will be stored in the system keyring):'
+                .format(self.username)
+            )
+            keyring.set_password(self.hostname, self.username, password)
+
+        return keyring.get_password(self.hostname, self.username)
+
+
+class AskPassword:
+
+    def __init__(self, hostname, username):
+        self.hostname = hostname
+        self.username = username
+
+    def get_password(self):
+        return getpass.getpass('Password for {}:'.format(self.username))
+
+
+def configure_providers(cfg):
+    # First, we'll try for the in-config one. It's not recommended, but someone
+    # may have set it.
+    try:
+        LOG.debug("Password provider in config as cleartext")
+    except KeyError:
```

Comment:
The log message isn't going to throw a key error, but the yield on line 53 might. I think this section needs to be reworked.

Reply:
*headdesk*